vbfml/scripts/preprocess_image.py

```diff
@@ -0,0 +1,437 @@
+
+    # load the tree and initilize batch info
+    Tree = uproot.open(input_files + ":sr_vbf")
+    batch_size = 5000
+    batch_number = ceil(Tree.num_entries / batch_size)
+    batch_counter = 0
+
+    # iterate by batches of 500 over the tree, aim to save memoryquit
+    for sub_tree in Tree.iterate(
+        ["JetImage_pixels", "leadak4_phi", "leadak4_eta"],
+        step_size=batch_size,
+        library="np",
+    ):
+
+        batch_counter += 1
+        print(f"batch {batch_counter}/{batch_number}")
+
+        batch_index = (batch_counter - 1) * batch_size
+
+        # initialize the new array for the preprocess images
+        New_images = np.zeros((batch_size, n_eta_bins * n_phi_bins), dtype="uint8")
+
+        # fix size of last batch
+        if batch_counter == batch_number:
+            batch_size = Tree.num_entries % batch_size
+            New_images = np.zeros((batch_size, n_eta_bins * n_phi_bins), dtype="uint8")
+
+        New_images_batch = np.ones((batch_size, n_eta_bins, n_phi_bins))
+        for j in range(batch_size):
+            New_images_batch[j] = sub_tree["JetImage_pixels"][j].reshape(im_shape)
+
+            # set phi = 0 (leading jet is at the center horizontal)
+            for i in range(n_eta_bins):
+                shift_phi = -round(
+                    sub_tree["leadak4_phi"][j] * n_phi_bins / (2 * np.pi)
+                )
+                New_images_batch[j][i] = np.roll(New_images_batch[j][i], shift_phi)
+
+            # set eta > 0 (leading jet is on the right side)
+            if sub_tree["leadak4_eta"][j] < 0:
+                New_images_batch[j] = np.flip(New_images_batch[j], 0)
+
+            New_images[j] = New_images_batch[j].flatten()
+
+        # writing in the new tree
+        new_tree = {}
+        new_tree["JetImage_pixels_preprocessed"] = New_images  # add the new branch
+        for branch in Tree.keys():  # copy all the other branches
+            new_tree[branch] = Tree[branch].arrays(
+                entry_start=batch_index,
+                entry_stop=batch_index + batch_size,
+                library="np",
+            )[branch]
+
+        if batch_counter == 1:
+            file["sr_vbf"] = new_tree
+        else:
+            file["sr_vbf"].extend(new_tree)
+
+        print(file["sr_vbf"].num_entries)
+
+
+@cli.command()
+@click.option(
+    "-i",
+    "--input-dir",
+    required=True,
+    help="Path to the directory with the input ROOT files.",
+)
+def rotate_all(input_dir: str):
+    """
+    Preprocess all root files of the directory by applying rotate function on every files of the input_dir
+    Warning -> have weird behavior over a lot of files. miss some events
+    """
+    files = glob.glob(input_dir + "/*root")
+
+    for file in files:
+        print(
+            "/////////////////// \nprocessing "
+            + os.path.basename(file)
+            + "\n///////////////////"
+        )
+        os.system("./preprocess_image.py rotate -i " + file)
```

Comment:
Okay but with this definition of the function then we will not be able to "rotate" only one root file right ?

Reply:
Sorry why do you think that? I was thinking that `rotate_all` command would be given the directory from command line, and it will call `rotate(file)` function for each `file` it finds within the directory. So the functionality should be the same, but I think it will be implemented in a cleaner way, unless I miss something.

Reply:
Ah wait I see your point now. To avoid that (and additional re-structuring of the code), then we can leave this code as it is now. So that if you only want to rotate a single file, you use the `rotate` command with just that file.

Reply:
yes I just wasn't sure how you implement it without removing the click fancy tools for the rotate function.

Reply:
okay good to know
I leave it like this for the moment then